sql/analyzer/resolve_columns.go

```diff
@@ -245,6 +245,26 @@ func qualifyColumns(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope, sel
 	// on the full scope; otherwise transform looks at sub-nodes and calculates a partial view of what is available.
 	symbols := getAvailableNamesByScope(n, scope)
 
+	var onDupUpdateSymbols availableNames
+	onDupUpdateLeftExprs := make(map[sql.Expression]bool)
+	if in, ok := n.(*plan.InsertInto); ok && len(in.OnDupExprs) > 0 {
+		inNoSrc := plan.NewInsertInto(
+			in.Database(),
+			in.Destination,
+			nil,
+			in.IsReplace,
+			in.ColumnNames,
+			in.OnDupExprs,
+			in.Ignore,
+		)
+		onDupUpdateSymbols = getAvailableNamesByScope(inNoSrc, scope)
+		for _, e := range in.OnDupExprs {
+			if sf, ok := e.(*expression.SetField); ok {
+				onDupUpdateLeftExprs[sf] = true
+			}
+		}
+	}
+
```

Comment:
Try pulling this out into a (hopefully documented) helper function that you could call from line 294. I think that'll keep the `qualifyColumns` more concise and readable and a good descriptive helper function name could also help document/explain what we're doing here to future code readers.

Reply:
I would want to call the helper method here, and not from line 294, to avoid repeating the same work right?

Reply:
It looks like the only reference this code block needs is `n`, right? When you're transforming on line 294, you are filtering to `InsertInto` nodes, so you'll have the node you need right there. I think the code would com out the cleanest if you just called that helper method from 294 to get the `onDupUpdateSymbols` that you need at that point in the code. Let me know if I'm just totally missing something though.

Reply:
The transform is a `transform.OneNodeExprsWithNode`, which traverses over the expressions returned from the `sql.Expressioner.Expression()` method. In the case of `sql.InsertInto`, it returns all the `OnDupExprs` and `Check` (`n` is still `InsertInto`, but `e` is a different expression each time), so it is definitely possible to run more than once. At least, I'm pretty sure that's what happens according to: https://github.com/dolthub/go-mysql-server/blob/main/sql/transform/node.go#L111

Reply:
sheet! you're right... I skimmed and thought we were transforming over nodes, not expressions. my bad. Pushing that code into a helper method still helps keep it tidy, but yeah... you're right – the current position is probably the right place to keep calling it so we don't recompute on every single expression.